index/scorch/segment/zap/docvalues.go

```diff
@@ -238,13 +252,31 @@ func (di *docValueReader) getDocValueLocs(docNum uint64) (uint64, uint64) {
 	return math.MaxUint64, math.MaxUint64
 }
 
+func (s *Segment) CurrentSegment() segment.DocumentFieldTermVisitable {
+	return s
+}
+
 // VisitDocumentFieldTerms is an implementation of the
 // DocumentFieldTermVisitable interface
-func (s *SegmentBase) VisitDocumentFieldTerms(localDocNum uint64, fields []string,
-	visitor index.DocumentFieldTermVisitor) error {
-	var dvIterClone *docValueReader
+func (s *Segment) VisitDocumentFieldTerms(localDocNum uint64, fields []string,
+	visitor index.DocumentFieldTermVisitor, dvsIn segment.DocVisitState) (
+	segment.DocVisitState, error) {
+	dvs, ok := dvsIn.(*docVisitState)
+	if !ok || dvs == nil {
+		dvs = &docVisitState{
+			dvrs:  make(map[uint16]*docValueReader, len(fields)),
+			state: &segment.FieldDocValueState{DvSegment: s},
+		}
+	} else {
+		if dvs.state.DvSegment != s {
+			dvs.state = &segment.FieldDocValueState{DvSegment: s}
+			dvs.dvrs = make(map[uint16]*docValueReader, len(fields))
+		}
+	}
+
 	fieldIDPlus1 := uint16(0)
-	ok := true
+	ok = true
```

Comment:
why is this needed? it seems like in all subsequent places we check it, we have assigned to it again